### DISCUSSION TITLE
Plugin architecture  = keep core simple

### DISCUSSION THREAD

**Discussion by jesperordrup** [plausible/analytics#1821] [Anything Else]
One way to keep plausible simple is by not adding more to core buy allow for plugins to solve different needs.

Exsmples for plugins:

Ecommerce
It's very different what ppl need..some wants funnels. Some just want to calculate checkouts and look at products and sales figures.

Visualisations
Some wants more maps others more graphs.

**Comment by metmarkosaric** [comment]
thanks for the feedback @jesperordrup! we do something like this with our script extensions. see https://plausible.io/docs/script-extensions. for the dashboard itself, we simply don't display the goals section for those that have no goals enabled. that seems like an approach to continue in the future too as we add funnels and other metrics. if you don't use it, it's not shown in the dashboard at all

**Comment by jesperordrup** [comment]
Thanks @metmarkosaric 

Yes I've seen it and it's cool. I understand your data collection approach and how it automagically is reflected in the UI. My suggestion was a bit further than that.

You cant make everybody happy and allow your product to grow out of its nice simplicity. But you can still offer a way for users to add functionality to their own installation.

Example:
A customer is interested in seing any data collected slightly different. It would make a huge different to the customer - rest of customers doesnt care. is it solvable? Yes - if you allow "plugin views".  With a plugin view i could show data differently, calculate, enrich. 

plugin views for one of my customers would be:

Top pages
- show full url

Top Source
- show full url, and some optional hover effects for showing additional info

Goals
- calcuale products sold with totals, categories sold totals, graph

Closing argument :-)

Its a way for a new product to please many and to stay clean. Otherwise the course is typically that you for a time is the new simple alternative - undtil product has maturied slowly with new features creaping in - until final state - enterprise - and then a new inititaive called "Refreshed Analytics" markteds itself as the simple alternative to the bloated 

I got carried away there - ill see myself out 😄   but theres a hint of truth in it, right?

**Reply by ukutaht** [reply]
Yes definitely. You're hitting on a lot of points that we are worried about ourselves. We want to satisfy as many use cases as possible but we don't want the product to become bloated.

In the grand scheme of things I see 3 directions:
1. Just design for the 80% use case and ignore requests for customization altogether. I think this is totally valid and reasonable. Plausible just doesn't show the full URL in Top Pages and that's that. The product is simple and inflexible, if it doesn't work for you then tough luck.
2. Some sort of plugin architecture like you explained. Inspiration from Matomo, Wordpress, etc
3. Add some of the most requested customization features into the settings. No plugins, everything is built-in.

This is an open question and I don't think we'll make a big decision on it in the short term. I will say this: I was against adding more settings so that we don't become GA 2.0. But this [blog post](https://linear.app/blog/settings-are-not-a-design-failure) has slightly changed my mind on the topic. Linear.app (product by the authors of the blog post) is absolutely perceived as a simple and streamlined product. So maybe it's OK to have a lot of settings as long as the product is designed well?

**Reply by jesperordrup** [reply]
If I think simplicity should win I'm mostly drawn towards 1 and/or 2. 

About 3)  I get the idea of a ton of settings could keep UI simple and only contain what ppl has configured. But it would still make the product itself complicated. And it would still require you to maintain codebase & PR's for a ton of feature requests and bugfixes. I think this would change your product and require more hands and it would mean that you would have to prroritize your time between a lot more issues - or teamMembers++

About 2)
I think that the plugin idea is much more valuable as it would allow ppl to get involved, customize and share solutions. It would open the box but still keep the solution simple and relevant to all. Plugins doesnt have to be Wordpress heavy (serverside and clientside). I would keep them "light weight" frontend plugins, aka js script that run in a frontend context with the data/endpoints that you offer.

About 1)
Better than 3.. and works. But everybody will forever miss 20%
